### ISSUE TITLE
WebpackError: ReferenceError: document is not defined

### ISSUE THREAD

**Issue by Huzi1** [AlexWang-16/react-lex-plus#9]
I am building my application using gatsby, everything runs fine during development mode but when I try to build it, I get this error 
![Screen Shot 2020-09-30 at 2 46 54 AM](https://user-images.githubusercontent.com/38772193/94588444-3a9c4780-02c7-11eb-8f2a-531fa1bf8565.png)
I am deploying my static website on Netlify, it pulls repo from git and builds it so even if I try some work around locally build will fail on Netlify.

**Comment by Huzi1** [comment]
This might help 
https://github.com/gatsbyjs/gatsby/issues/19487

**Comment by AlexWang-16** [comment]
Thanks for submitting this bug and the information @Huzi1. I'll investigate the issue.

**Comment by AlexWang-16** [comment]
Hi @Huzi1, I've looked into the issue and it seems to me that it is more complicated than I had expected. I'll leave you with 2 versions of my answer.

# TLDR Answer
While react supports server side rendering, it is by nature client side rendering oriented. The errors were thrown as soon as react-lex-plus project is imported via `import LexChat from "react-lex-plus"` as it executes code from its compiled version. 

This rules out the first suggestion of adding checks for `document`. Adding checks in the project itself also did not yield desirable results. React's job is to take the JSX written and converting it into HTML and manipulating the document object model of the browser directly.

The best solution is to use [loadable-components](https://github.com/gregberge/loadable-components), which I've tested to be working 100% with react-lex-plus. I would recommend using this method as the best solution.

Here's a sample code that I've used to get it working, which you can use as reference. Be sure to run `npm i @loadable/component` first:
```
import React from "react"
import loadable from "@loadable/component"

const LexChat = loadable(() => import("react-lex-plus"));
export default function Home() {
  return (
  <>
    <LexChat
      botName="HelloBot"
      IdentityPoolId="us-east-1:e464decf-XXXX-XXXX-XXXX-XXXXXXXX"
      placeholder="Enter Text"
      backgroundColor="#FFFFFF"
      height="430px"
      region="us-east-1"
      headerText="Assistant"
      headerStyle={{ backgroundColor: "#ABD5D9", fontSize: "30px" }}
      greeting={
        "Hello, how can I help? You can say things like 'help' to get more info"
      }
    />
  </>
  )
}
```

# Long Answer

I'm going to be upfront and honest in admitting that my knowledge of how React works internally is limited. When I attempted to address the issue by adding if statements to check if document has been loaded in the lifecycle methods as well as the component's constructor, Gatsby would still throw the same error. It looks like when I built react-lex-plus, webpack transpiled the code using React, which invoked calls against `document`, which is out of my control.

During the troubleshooting and development process, Gatsby wasn't playing nicely with my development version of react-lex-plus despite both having the same version of webpack and babel. It will throw syntax errors when I was running `gatsby develop` while it compiled perfectly for the react-lex-plus project. This I still cannot understand and hindered my testing process.

I would suggest referring to @wardpeet's suggestion of using loadable-components, which is a wrapper that will delay loading the component until `document` is available, which I've tested to work with Gatsby.

Here's a sample code that I've used to get it working, which you can use as reference. Be sure to run `npm i @loadable/component` first:
```
import React from "react"
import loadable from "@loadable/component"

const LexChat = loadable(() => import("react-lex-plus"));
export default function Home() {
  return (
  <>
    <LexChat
      botName="HelloBot"
      IdentityPoolId="us-east-1:e464decf-XXXX-XXXX-XXXX-XXXXXXXX"
      placeholder="Enter Text"
      backgroundColor="#FFFFFF"
      height="430px"
      region="us-east-1"
      headerText="Assistant"
      headerStyle={{ backgroundColor: "#ABD5D9", fontSize: "30px" }}
      greeting={
        "Hello, how can I help? You can say things like 'help' to get more info"
      }
    />
  </>
  )
}
```

While this may not be the answer you're hoping for, I hope this would at least provide you with a solution around the problem you're having. Feel free to leave questions or any comments if there's something I did not explain clearly.

**Comment by Huzi1** [comment]
Hi @AlexWang-16 ,
Thank you for your detail response. I figured it out three days ago, I used the same loaded component package.I totally forgot to post it here.
You are correct by saying that react by nature is client side rendering. Gatsby deployment is SSR(server side rendering). Here is the final result (https://www.yaminhuzaifa.xyz/).
